File: c_ver/Complex.c
#include "Complex.h"

#include <math.h>

Complex addComplex(Complex a, Complex b) {
    Complex result;
    result.real = a.real + b.real;
    result.imag = a.imag + b.imag;
    return result;
}

Complex subComplex(Complex a, Complex b) {
    Complex result;
    result.real = a.real - b.real;
    result.imag = a.imag - b.imag;
    return result;
}

Complex mulComplex(Complex a, Complex b) {
    Complex result;
    result.real = a.real * b.real - a.imag * b.imag;
    result.imag = a.real * b.imag + a.imag * b.real;
    return result;
}

Complex divComplex(Complex a, Complex b) {
    Complex result;
    double denominator = b.real * b.real + b.imag * b.imag;
    result.real = (a.real * b.real + a.imag * b.imag) / denominator;
    result.imag = (a.imag * b.real - a.real * b.imag) / denominator;
    return result;
}

Complex newComplexDescartes(double real, double imag) {
    Complex c;
    c.real = real;
    c.imag = imag;

    return c;
}

Complex newComplexPolar(double magnitude, double angle) {
    Complex c;
    c.real = magnitude * cos(angle);
    c.imag = magnitude * sin(angle);

    return c;
}

File: rust_ver/my_fft_rustver/src/Complex.rs
#[derive(Clone, Copy)]
pub struct myComplex {
    pub re: f64,
    pub im: f64,
}

impl myComplex {
    pub fn newDescartes(re: f64, im: f64) -> Self {
        myComplex { re, im }
    }

    pub fn newPolar(r: f64, theta: f64) -> Self {
        myComplex {
            re: r * theta.cos(),
            im: r * theta.sin(),
        }
    }
}

pub fn addComplex(a: &myComplex, b: &myComplex) -> myComplex {
    myComplex {
        re: a.re + b.re,
        im: a.im + b.im,
    }
}

pub fn subComplex(a: &myComplex, b: &myComplex) -> myComplex {
    myComplex {
        re: a.re - b.re,
        im: a.im - b.im,
    }
}

pub fn mulComplex(a: &myComplex, b: &myComplex) -> myComplex {
    myComplex {
        re: a.re * b.re - a.im * b.im,
        im: a.re * b.im + a.im * b.re,
    }
}

pub fn divComplex(a: &myComplex, b: &myComplex) -> myComplex {
    let denom = b.re * b.re + b.im * b.im;
    myComplex {
        re: (a.re * b.re + a.im * b.im) / denom,
        im: (a.im * b.re - a.re * b.im) / denom,
    }
}